cs3api4lab/cs3apismanager.py

```diff
@@ -57,15 +140,374 @@ def save(self, model, path):
+            model = self._notebook_model(path, content=False)
+
+        elif model['type'] == 'file':
+            model = self._file_model(path, content=False, format=None)
+
+        elif model['type'] == 'directory':
+            model = self._dir_model(path, content=False)
+
+        if validation_message:
+            model['message'] = validation_message
+
+        # self.run_post_save_hook(model=model, os_path=path)
+
+        return model
 
     def delete_file(self, path):
         """Delete the file or directory at path."""
-        raise NotImplementedError('cs3: missing')
+
+        path = self._normalize_path(path)
+
+        try:
+            self._cs3_file_api().remove(path, self.cs3_user_id, self.cs3_endpoint)
+        except Exception as e:
+            self.log.error(u'Unknown error delete file: %s %s', path, e, exc_info=True)
+            raise web.HTTPError(500, u'Unknown error delete file: %s %s' % (path, e))
 
     def rename_file(self, old_path, new_path):
         """Rename a file or directory."""
-        raise NotImplementedError('cs3: missing')
 
-    
+        if new_path == old_path:
+            return
+
+        #
+        # ToDo: Implements validate file like: notebook/services/contents/filemanager.py:587 using Reva API
+        #
+        old_path = self._normalize_path(old_path)
+        new_path = self._normalize_path(new_path)
+
+        # Move the file
+        try:
+            cs3_file_api = self._cs3_file_api()
+            cs3_file_api.move(old_path, new_path, self.cs3_user_id, self.cs3_endpoint)
+        except Exception as e:
+            self.log.error(u'Error renaming file: %s %s', old_path, e)
+            raise web.HTTPError(500, u'Error renaming file: %s %s' % (old_path, e))
+
+    def new(self, model=None, path=''):
+
+        path = path.strip('/')
+        if model is None:
+            model = {}
+
+        if path.endswith('.ipynb'):
+            model.setdefault('type', 'notebook')
+        else:
+            model.setdefault('type', 'file')
+
+        # no content, not a directory, so fill out new-file model
+        if 'content' not in model and model['type'] != 'directory':
+            if model['type'] == 'notebook':
+                model['content'] = new_notebook()
+                model['format'] = 'json'
+            else:
+                model['content'] = ''
+                model['type'] = 'file'
+                model['format'] = 'text'
+
+        model = self.save(model, path)
+
+        # ToDo: Fix writable flag - based on container status
+        model['writable'] = True
+
+        return model
+
+    def _cs3_file_api(self):
+        return Cs3FileApi(self.cs3_config, self.log)
+
+    def _normalize_path(self, path):
+
+        if len(path) > 0 and path[0] != '/':
+            path = '/' + path
+        elif path == '' or path is None:
+            path = '/'
+        return path
+
+    def _get_parent_path(self, path):
```

Comment:
I tried to use path form OS or pathlib (something like). I remember it was not working properly because I'm don't use real filesystem.

Reply:
It works with strings. It shouldn't matter wether it's a real filesystem or not.

Reply:
Yes, pathlib works with strings, but 'path.parent' it's not string, this is an object. It's don't have methods form string type. 
When I use "Pathlib" throws exception: "TypeError: object of type 'WindowsPath' has no len()"

Reply:
I still don't understand. Everywhere I look it seems to me that you are passing strings to `_get_parent_path`. Is there's any place where it isn't, it could be converted for that particular case. But I don't want to delay this PR for much longer, so I'll ignore it for now and create an issue to check after.